### DISCUSSION TITLE
Microservices Roadmap

### DISCUSSION THREAD

**Discussion by ukaoskid** [pequehq/framework#89] [Ideas]
# Abstract

The idea is to extract the Microservices layer from the Framework package and move it into a separate package.
In my humble opinion, whether to put the Microservice package inside the framework monorepo or into a new dedicated repo is still to be well understood, but for the sake of the initial proof-of-concept, we chose to continue in the same Framework monorepo.

# Food for thought

In order to make an informed decision about whether to keep the Microservices package inside the Framework monorepo or to make it a dedicated repo we have to consider the following questions.

| Question | Possible answer, but a topic for discussion |
|----------|------------------------------------------|
| Are we going to make this set of services usable out of Peque? | It's yes and no, in principle I would say no, but there could be a certain degree of exception, like the `gRPC` service transposer (more similar to the `GraphQL` resolver transposer). |
| Hybrid solution? Broker-based microservices in and gRPC out | This could be a good solution, indeed, all the MQTT, Redis, Kafka, and SMB services won't be used outside of Peque, but gRPC that could be. |

## Broker based services vs. gRPC

There is a substantial difference because, in fact, they are two different design patterns (broker and client/server) and they serve two different purposes.

### Broker

The initial (primordial) implementation of all the broker's types (MQTT, Redis, Kafka, and SMB) are already reflecting the baseline concept of "transposing into OOP". The `@ConsumeEvent` and `@ProduceEvent` are already driving the developers to the usage of a class-pattern strategy as they are tied to the `@Microservice` class decorator.

### Client/Server

The same concept shall be applied to this pattern as well. At this moment, the only client/server pattern we want to implement is the `gRPC` one.

### Server

The implementation idea for the gRPC is absolutely similar to the GraphQL package, which is creating a set of decorators that shall generate the needed metadata to create the `Server` in an OOP fashion (that would respect the baseline concept of "transposing into OOP").

### Client

For the client, there isn't much to do as the services are inherited by the proto files directly and they can be implemented easily.
The only component that shall be exported is then a `ClientModule` that will accept some configurations such as `proto`, `credentials`, and `destination server`.

## Conclusion

In conclusion, I would say that the optimal solution is to divide the Microservices package like the following.

| Feature | Sub-features | Repository |
|---------|--------------|------------|
| MQTT | | Framework/Microservices |
| Redis | | Framework/Microservices |
| Kafka | | Framework/Microservices |
| SMB | | Framework/Microservices |
| gRPC | | gRPC |

The `gRPC` will be the only service that could be used separately from the framework as it works as an OOP pattern facilitator, exactly like the `GraphQL` library.

**Comment by alesmit** [comment]
I agree that as long as gRPC decorators are not tied to the framework they can live in a separate package and repository the same as it has been done with `@pequehq/graphql`. Among the other things, monorepos are meant to facilitate the management of dependencies within the same ecosystem, but since gRPC decorators are not a dependency of the framework and, vice-versa, the framework is not a dependency of them, I strongly support the idea of keeping them in a separate repository with their own versioning and release process.

The packages we'll build to support microservices such as MQTT, Redis, etc. are different because their goal is to unlock a functionality within the framework, which already encourages OOP patterns. From how I imagine them to work, it shouldn't make sense to install microservices packages without the framework because it already has decorators and logic in place to handle events, so the functions or decorators exported by these packages would only wrap the SDK of X message broker (Kafka, Redis, etc.) and act as a bridge between their API and the framework API, providing a seamless implementation.
If, instead, our goal is only to encourage OOP patterns while using, for instance, Kafka outside of the framework, those packages should not be tied to the framework. I advocate against this idea, the reason being that the effort to install the SDK and make your own OOP implementation of the Redis client is minimal compared to what it takes to achieve the same result with gRPC or GraphQL, which have fairly complex structure and thus it makes sense to simplify their use via decorators.

**Reply by ukaoskid** [reply]
Then it looks like we agree on the following line of strategy:

| Package | Content | Repository |
|----------|---------|------------|
| gRPC | gRPC Server decorators and ClientModule | gRPC |
| Brokers | MQTT, Redis, Kafka, SMB | Framework/Microservices |
